File: PunctualPath/TimetableData-PyCharm/ImageProcessing.py
import colorsys

import cv2
import numpy as np
from PIL import Image
from matplotlib import pyplot as plt
from ocrmac import ocrmac

from utils import TimeTable, Row


def is_color_in_range(r, g, b):
    h, s, v = colorsys.rgb_to_hsv(r, g, b)
    return 0.55 <= h <= 0.7 and 0.01 <= s and 65 <= v <= 240


def increase_contrast(image: Image.Image):
    ret, th = cv2.threshold(np.array(image.convert('L')),
                            0,  # threshold value, ignored when using cv2.THRESH_OTSU
                            255,  # maximum value assigned to pixel values exceeding the threshold
                            cv2.THRESH_BINARY + cv2.THRESH_OTSU)
    return Image.fromarray(th)


def filter_image(img: Image.Image):
    img = img.convert("RGB")  # Ensure image is in RGB format
    pixels = img.load()  # Create the pixel map

    # Get image dimensions
    width, height = img.size

    # Iterate over all pixels
    for y in range(height):
        for x in range(width):
            # Get the RGBA value of the pixel
            r, g, b = pixels[x, y]

            if not is_color_in_range(r, g, b):
                pixels[x, y] = (255, 255, 255)

    # new image
    return img


def locate_nth_color_change(img: Image.Image,
                            n: int,
                            vertical: bool,
                            start: tuple,
                            step: int = 1,
                            skip: int = 1,
                            skip_x: int = 0,
                            skip_y: int = 0,
                            percentage=False,
                            tolerance_high=100,
                            tolerance_low=5,
                            annotate=False) -> tuple:
    width, height = img.size

    if isinstance(start[0], int):
        x, y = start
    else:
        x, y = int(start[0] * width), int(start[1] * height)

    if skip_x == 0 and skip_y == 0:
        if vertical:
            y += skip
        else:
            x += skip
    else:
        x += skip_x
        y += skip_y

    r_prev, g_prev, b_prev = img.getpixel(xy=(x, y))
    color_change_count = 0
    changing = False
    while color_change_count < n and 0 < x < width - step and 0 < y < height - step:
        if vertical:
            y += step
        else:
            x += step
        r, g, b = img.getpixel(xy=(x, y))
        rgb_change = abs(r - r_prev) + abs(g - g_prev) + abs(b - b_prev)
        if rgb_change > tolerance_high and changing is False:
            color_change_count += 1
            changing = True
            if annotate:
                img.putpixel(xy=(x - 3, y - 3), value=(0, 0, 0))
                img.putpixel(xy=(x - 3, y + 3), value=(0, 0, 0))
                img.putpixel(xy=(x + 3, y - 3), value=(0, 0, 0))
                img.putpixel(xy=(x + 3, y + 3), value=(0, 0, 0))
        elif rgb_change < tolerance_low and changing is True:
            changing = False
        r_prev, g_prev, b_prev = r, g, b

    if percentage:
        return x / width, y / height
    return x, y


def locate_hour_strip(img: Image.Image, verbose=False) -> tuple:
    left_0 = locate_nth_color_change(img, 2, vertical=False, start=(0.02, 0.4), tolerance_high=50)
    left = locate_nth_color_change(img, 1, vertical=True, start=left_0, skip_x=1, skip_y=1)
    right = locate_nth_color_change(img, 1, vertical=False, start=left)
    if verbose:
        print("Hour Strip", left, right)
    return left, right


def locate_row_positions(img: Image.Image, hour_strip_right: tuple, verbose=False) -> list[int]:
    hour_strip_top_right = locate_nth_color_change(img, 1, vertical=True, start=hour_strip_right, step=-1, skip_x=8,
                                                   tolerance_high=100, annotate=True)
    row_1_b = locate_nth_color_change(img, 1, vertical=True, start=hour_strip_top_right, step=1, skip_x=1, skip_y=5,
                                      tolerance_high=20, annotate=True)
    row_height = row_1_b[1] - hour_strip_top_right[1]
    if verbose:
        print("Hour Strip Top Right", hour_strip_top_right)
        print("Row Width", row_height)
    row_b = row_1_b

    rows: list[int] = [hour_strip_top_right[1], row_1_b[1]]

    while row_b[1] < img.size[1] - row_height * 1.5:
        row_b_r = locate_nth_color_change(img, 1, vertical=True, start=row_b, step=1, skip_y=int(row_height * 0.9),
                                          annotate=True, tolerance_high=20)
        if row_b_r[1] - rows[-1] > row_height * 1.5:
            row_b_l = locate_nth_color_change(img, 1, vertical=True, start=row_b, step=1, skip_x=-12,
                                              skip_y=int(row_height * 0.9), annotate=True, tolerance_high=50)
            if row_b_l[1] - rows[-1] > row_height * 1.5:
                break
            else:
                row_b = (row_b_l[0] + 12, row_b_l[1])
        else:
            row_b = row_b_r
        rows.append(row_b[1])

    return rows


def load_timetable(img: Image.Image, filter=True, verbose=False) -> TimeTable:
    hour_strip_l, hour_strip_r = locate_hour_strip(img, verbose=verbose)
    rows = locate_row_positions(img, hour_strip_r, verbose=verbose)

    table = TimeTable()
    h = 4
    for i in range(len(rows) - 1):
        x_l = hour_strip_l[0]
        x_m = hour_strip_r[0]
        x_r = img.size[0]
        y_t = rows[i]
        y_b = rows[i + 1]
        # y_m = (y_b + y_t) // 2
        # height = y_b - y_t

        h_crop = img.crop((x_l, y_t, x_m, y_b))
        if filter:
            h_crop = filter_image(h_crop)
            h_crop = increase_contrast(h_crop)
        if verbose:
            h_crop.show()

        try:
            h_raw = int(ocrmac.OCR(h_crop).recognize()[0][0])
            if h_raw != h + 1 and h != 23:
                if len(table.rows) >= 3 and table.rows[-1].hour == table.rows[-2].hour + 1 and table.rows[-1].hour == \
                        table.rows[-3].hour + 2:
                    h += 1
                    if h == 24:
                        h = 0
                    print(f"Hour recognized incorrectly at {h} ({h_raw})")
            else:
                h = h_raw
        except Exception as e:
            h += 1
            if h == 24:
                h = 0
            print(f"Hour not recognized at {h}")

        row = Row(h)

        m_crop = img.crop((x_m, y_t, x_r, y_b))
        if filter:
            m_crop = filter_image(m_crop)
            m_crop = increase_contrast(m_crop)
        if verbose:
            m_crop.show()
        # x_running = 0
        # dist = []
        # while x_running < m_crop.size[0] - x_l:
        #     x_prev = x_running
        #     x_running, _ = locate_nth_color_change(m_crop, 1, vertical=False, start=(x_running, height // 2), annotate=True)
        #     dist.append(x_running - x_prev)
        # if h == 18:
        # m_crop.show()

        m_result = ocrmac.OCR(m_crop).recognize()
        m = ""
        for result in m_result:
            m += result[0] + ' '
        m = m.strip()
        m_formatted = format_minutes(m)
        row.minutes = m_formatted
        table.add_row(row)
    return table


def format_minutes(m: str) -> list[int]:
    m = m.replace('Z', '2').replace('S', '5')
    m = ''.join(c if c.isdigit() else ' ' for c in m).replace("  ", " ")
    m_list = m.split(' ')
    new_list = []
    for i, num_str in enumerate(m_list):
        if len(num_str) == 0:
            continue

        num_int = int(num_str)

        if num_int > 59 and num_int < 1000:
            rm_r = int(num_str[0:2])
            rm_l = int(num_str[1:3])

            if i == 0:
                prior_m = -1
            else:
                prior_m = int(m_list[i - 1])

            if i == len(m_list) - 1:
                next_m = 60
            else:
                try:
                    next_m = int(m_list[i + 1])
                except ValueError:
                    next_m = prior_m + 10
                    print(f"Inaccurate result from parsing {num_str}")

            midpoint = (next_m - prior_m) / 2

            if not prior_m < rm_r < next_m and prior_m < rm_l < next_m:
                num_int = rm_l
            elif not prior_m < rm_l < next_m and prior_m < rm_r < next_m:
                num_int = rm_r
            elif prior_m < rm_r < next_m and prior_m < rm_l < next_m:
                if abs(rm_r - midpoint) < abs(rm_l - midpoint):
                    num_int = rm_r
                else:
                    num_int = rm_l
            else:
                num_int = midpoint
                print("Failed to parse minute data {}, using midpoint value {}".format(num_str, midpoint))

        if num_int not in new_list:
            new_list.append(num_int)

    return new_list


def get_hsv_of_img(img: Image):
    h = []
    s = []
    v = []
    for x in range(img.size[0]):
        for y in range(img.size[1]):
            r, g, b, a = img.getpixel((x, y))

            if a == 255:
                h_, s_, v_ = colorsys.rgb_to_hsv(r, g, b)
                h.append(h_)
                s.append(s_)
                v.append(v_ / 255)

    plt.boxplot([h, s, v])
    plt.show()

File: _site/PunctualPath/ImageProcessing.swift
//
//  ImageProcessing.swift
//  PunctualPath
//
//  Created by Roy Zhao on 7/8/24.
//

import Foundation
import UIKit
import Vision

func recognizeText(from image: UIImage, completion: @escaping ([String]) -> Void) {
    guard let cgImage = image.cgImage else {
        completion([])
        return
    }
    
    let requestHandler = VNImageRequestHandler(cgImage: cgImage, options: [:])
    let request = VNRecognizeTextRequest { (request, error) in
        guard let observations = request.results as? [VNRecognizedTextObservation] else {
            completion([])
            return
        }
        
        let recognizedStrings = observations.compactMap { observation in
            return observation.topCandidates(1).first?.string
        }
        
        completion(recognizedStrings)
    }
    
    request.recognitionLevel = .accurate
    
    do {
        try requestHandler.perform([request])
    } catch {
        print("Error performing text recognition: \(error)")
        completion([])
    }
}

func isColorInRange(r: Int, g: Int, b: Int) -> Bool {
    var h: CGFloat = 0, s: CGFloat = 0, v: CGFloat = 0
    UIColor(red: CGFloat(r) / 255.0, green: CGFloat(g) / 255.0, blue: CGFloat(b) / 255.0, alpha: 1.0).getHue(&h, saturation: &s, brightness: &v, alpha: nil)
    return (0.55...0.7).contains(h) && (0.01...1.0).contains(s) && (65...240).contains(Int(v * 255))
}

func increaseContrast(image: UIImage) -> UIImage? {
    guard let ciImage = CIImage(image: image) else { return nil }
    let filter = CIFilter(name: "CIColorControls")
    filter?.setValue(ciImage, forKey: kCIInputImageKey)
    filter?.setValue(2.0, forKey: kCIInputContrastKey) // Increase contrast
    guard let outputCIImage = filter?.outputImage else { return nil }
    return UIImage(ciImage: outputCIImage)
}

func filterImage(img: UIImage) -> UIImage? {
    guard let cgImage = img.cgImage else { return nil }
    let colorSpace = CGColorSpaceCreateDeviceRGB()
    let width = cgImage.width
    let height = cgImage.height
    let bytesPerPixel = 4
    let bytesPerRow = bytesPerPixel * width
    let bitsPerComponent = 8
    var pixelData = [UInt8](repeating: 0, count: width * height * bytesPerPixel)
    
    let context = CGContext(data: &pixelData,
                            width: width,
                            height: height,
                            bitsPerComponent: bitsPerComponent,
                            bytesPerRow: bytesPerRow,
                            space: colorSpace,
                            bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue | CGBitmapInfo.byteOrder32Big.rawValue)
    
    guard let context = context else { return nil }
    context.draw(cgImage, in: CGRect(x: 0, y: 0, width: width, height: height))
    
    for y in 0..<height {
        for x in 0..<width {
            let offset = (y * width + x) * bytesPerPixel
            let r = CGFloat(pixelData[offset]) / 255.0
            let g = CGFloat(pixelData[offset + 1]) / 255.0
            let b = CGFloat(pixelData[offset + 2]) / 255.0
            
            if !isColorInRange(r: Int(r * 255), g: Int(g * 255), b: Int(b * 255)) {
                pixelData[offset] = 255
                pixelData[offset + 1] = 255
                pixelData[offset + 2] = 255
            }
        }
    }
    
    guard let outputCGImage = context.makeImage() else { return nil }
    return UIImage(cgImage: outputCGImage)
}

func locateNthColorChange(img: UIImage,
                          n: Int,
                          vertical: Bool,
                          start: CGPoint,
                          step: Int = 1,
                          skip: Int = 1,
                          skipX: Int = 0,
                          skipY: Int = 0,
                          percentage: Bool = false,
                          toleranceHigh: Int = 100,
                          toleranceLow: Int = 5,
                          annotate: Bool = false) -> CGPoint {
    
    guard let cgImage = img.cgImage else {
        fatalError("Failed to get CGImage from UIImage")
    }
    
    let width = cgImage.width
    let height = cgImage.height
    var x: Int = 0
    var y: Int = 0
    if start.x < 1 {
        x = Int(start.x * CGFloat(width))
    } else {
        x = Int(start.x)
    }
    
    if start.y < 1 {
        y = Int(start.y * CGFloat(height))
    } else {
        y = Int(start.y)
    }
    
    
    if skipX == 0 && skipY == 0 {
        if vertical {
            y += skip
        } else {
            x += skip
        }
    } else {
        x += skipX
        y += skipY
    }
    
    let colorSpace = CGColorSpaceCreateDeviceRGB()
    let bytesPerPixel = 4
    let bytesPerRow = bytesPerPixel * width
    let bitsPerComponent = 8
    var pixelData = [UInt8](repeating: 0, count: width * height * bytesPerPixel)
    
    let context = CGContext(data: &pixelData,
                            width: width,
                            height: height,
                            bitsPerComponent: bitsPerComponent,
                            bytesPerRow: bytesPerRow,
                            space: colorSpace,
                            bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue | CGBitmapInfo.byteOrder32Big.rawValue)
    
    guard let context = context else {
        fatalError("Failed to create CGContext")
    }
    
    context.draw(cgImage, in: CGRect(x: 0, y: 0, width: width, height: height))
    
    let offsetInitial = (y * width + x) * bytesPerPixel
    var rPrev: Int = Int(pixelData[offsetInitial]), gPrev: Int = Int(pixelData[offsetInitial + 1]), bPrev: Int = Int(pixelData[offsetInitial + 2])
    var colorChangeCount = 0
    var changing = false
    
    while colorChangeCount < n && 0 < x && x < width - step && 0 < y && y < height - step {
        if vertical {
            y += step
        } else {
            x += step
        }
        
        let offset = (y * width + x) * bytesPerPixel
        let r = Int(pixelData[offset])
        let g = Int(pixelData[offset + 1])
        let b = Int(pixelData[offset + 2])
        
        let rgbChange = abs(r - rPrev) + abs(g - gPrev) + abs(b - bPrev)
        
        if rgbChange > toleranceHigh && !changing {
            colorChangeCount += 1
            changing = true
            
            if annotate {
                // Assuming annotating by setting pixels to black
                let annotateRadius = 3
                for dx in [-annotateRadius, annotateRadius] {
                    for dy in [-annotateRadius, annotateRadius] {
                        let ax = x + dx
                        let ay = y + dy
                        if ax >= 0 && ax < width && ay >= 0 && ay < height {
                            let annotateOffset = (ay * width + ax) * bytesPerPixel
                            pixelData[annotateOffset] = 0 // Red component
                            pixelData[annotateOffset + 1] = 0 // Green component
                            pixelData[annotateOffset + 2] = 0 // Blue component
                        }
                    }
                }
            }
        } else if rgbChange < toleranceLow && changing {
            changing = false
        }
        
        rPrev = r
        gPrev = g
        bPrev = b
    }
    
    if percentage {
        return CGPoint(x: CGFloat(x) / CGFloat(width), y: CGFloat(y) / CGFloat(height))
    } else {
        return CGPoint(x: CGFloat(x), y: CGFloat(y))
    }
}

func locateHourStrip(img: UIImage, verbose: Bool = false) -> (CGPoint, CGPoint) {
    let left0 = locateNthColorChange(img: img, n: 2, vertical: false, start: CGPoint(x: 0.02, y: 0.4), toleranceHigh: 50)
    let left = locateNthColorChange(img: img, n: 1, vertical: true, start: left0, skipX: 1, skipY: 1)
    let right = locateNthColorChange(img: img, n: 1, vertical: false, start: left)
    
    if verbose {
        print("Hour Strip", left, right)
    }
    
    return (left, right)
}

func locateRowPositions(img: UIImage, hourStripRight: CGPoint, verbose: Bool = false) -> [Int] {
    let hourStripTopRight = locateNthColorChange(img: img, n: 1, vertical: true, start: hourStripRight, step: -1, skipX: 8, toleranceHigh: 100, annotate: true)
    
    let row1B = locateNthColorChange(img: img, n: 1, vertical: true, start: hourStripTopRight, step: 1, skipX: 1, skipY: 5, toleranceHigh: 20, annotate: true)
    let rowHeight = row1B.y - hourStripTopRight.y
    
    if verbose {
        print("Hour Strip Top Right", hourStripTopRight)
        print("Row Width", rowHeight)
    }
    
    var rowB = row1B
    var rows = [hourStripTopRight.y, row1B.y]
    
    while rowB.y < img.size.height - CGFloat(rowHeight) * 1.5 {
        let rowBR = locateNthColorChange(img: img, n: 1, vertical: true, start: rowB, step: 1, skipY: Int(rowHeight * 0.9), toleranceHigh: 20, annotate: true)
        
        if rowBR.y - rows.last! > CGFloat(rowHeight) * 1.5 {
            let rowBL = locateNthColorChange(img: img, n: 1, vertical: true, start: rowB, step: 1, skipX: -12, skipY: Int(rowHeight * 0.9), toleranceHigh: 50, annotate: true)
            
            if rowBL.y - rows.last! > CGFloat(rowHeight) * 1.5 {
                break
            } else {
                rowB = CGPoint(x: rowBL.x + 12, y: rowBL.y)
            }
        } else {
            rowB = rowBR
        }
        rows.append(rowB.y)
    }
    
    return rows.map { Int($0) }
}

func replaceNonDigitsWithSingleSpace(in input: String) -> String {
    let pattern = "\\D+"
    let regex = try! NSRegularExpression(pattern: pattern, options: [])
    let range = NSRange(location: 0, length: input.utf16.count)
    return regex.stringByReplacingMatches(in: input, options: [], range: range, withTemplate: " ")
}

func formatMinutes(_ m: String) -> [Int] {
    var m = m.replacingOccurrences(of: "Z", with: "2").replacingOccurrences(of: "S", with: "5")
    m = replaceNonDigitsWithSingleSpace(in: m)
    let mList = m.split(separator: " ").map { String($0) }
    var newList: [Int] = []

    for (i, numStr) in mList.enumerated() {
        if numStr.isEmpty {
            continue
        }

        guard var numInt = Int(numStr) else { continue }

        if numInt > 59 && numInt < 1000 {
            let rmR = Int(numStr.prefix(2)) ?? 0
            let rmL = Int(numStr.suffix(2)) ?? 0

            let priorM: Int
            if i == 0 {
                priorM = -1
            } else {
                priorM = Int(mList[i - 1]) ?? 0
            }

            let nextM: Int
            if i == mList.count - 1 {
                nextM = 60
            } else {
                nextM = Int(mList[i + 1]) ?? priorM + 10
            }

            let midpoint = (nextM - priorM) / 2

            if !(priorM..<nextM).contains(rmR), (priorM..<nextM).contains(rmL) {
                numInt = rmL
            } else if !(priorM..<nextM).contains(rmL), (priorM..<nextM).contains(rmR) {
                numInt = rmR
            } else if (priorM..<nextM).contains(rmR), (priorM..<nextM).contains(rmL) {
                if abs(rmR - midpoint) < abs(rmL - midpoint) {
                    numInt = rmR
                } else {
                    numInt = rmL
                }
            } else {
                numInt = midpoint
                print("Failed to parse minute data \(numStr), using midpoint value \(midpoint)")
            }
        }

        if !newList.contains(numInt) {
            newList.append(numInt)
        }
    }

    return newList
}

func loadTimetable(img: UIImage, filter: Bool = false, verbose: Bool = false) -> TimeTable {
    let (hourStripLeft, hourStripRight) = locateHourStrip(img: img, verbose: verbose)
    let rowPositions = locateRowPositions(img: img, hourStripRight: hourStripRight, verbose: verbose)
    
    let table = TimeTable()
    var hour = 4
    
    for i in 0..<rowPositions.count - 1 {
        let xLeft = Int(hourStripLeft.x)
        let xMiddle = Int(hourStripRight.x)
        let xRight = img.size.width
        let yTop = rowPositions[i]
        let yBottom = rowPositions[i + 1]
        
        let hCropRect = CGRect(x: xLeft, y: yTop, width: xMiddle - xLeft, height: yBottom - yTop)
        
        var hCrop = img.cropped(to: hCropRect)
            
        if filter {
            hCrop = filterImage(img: hCrop!)
            hCrop = increaseContrast(image: hCrop!)
        }
//        if verbose {
            // Display cropped and processed image
            // Assuming you have a method to display UIImage
//        displayImage(image: hCrop)
//        }
        
        var hRaw = 0
        
        do {
            if let image = hCrop {
                recognizeText(from: image) { ocrResult in
                    print("Hour")
                    print(ocrResult)
                    hRaw = Int(ocrResult[0]) ?? 0
                    
                    if hRaw != hour + 1 && hour != 23 {
                        if table.rows.count >= 3 && table.rows[table.rows.count - 1].hour == table.rows[table.rows.count - 2].hour + 1 && table.rows[table.rows.count - 1].hour == table.rows[table.rows.count - 3].hour + 2 {
                            hour += 1
                            if hour == 24 {
                                hour = 0
                            }
                            print("Hour recognized incorrectly at \(hour) (\(hRaw))")
                        }
                    } else {
                        hour = hRaw
                    }
                }
                
            }
        } catch {
            hour += 1
            if hour == 24 {
                hour = 0
            }
            print("Hour not recognized at \(hour)")
        }
        
        let row = Row(hour: hour)
        
        let mCropRect = CGRect(x: xMiddle, y: yTop, width: Int(xRight) - xMiddle, height: yBottom - yTop)
        var mCrop = img.cropped(to: mCropRect)
        
        if filter {
            mCrop = filterImage(img: mCrop!)
            mCrop = increaseContrast(image: mCrop!)
        }
        
//        if verbose {
//            // Display cropped and processed image
//            // Assuming you have a method to display UIImage
//            displayImage(image: mCrop)
//        }
        
        do {
            if let image = mCrop {
                var minutes = ""
                recognizeText(from: image) { ocrResult in
                    for result in ocrResult {
                        print("Min")
                        print(result)
                        minutes += result + " "
                    }
                }
                                
                minutes = minutes.trimmingCharacters(in: .whitespaces)
//                
                let formattedMinutes = formatMinutes(minutes)
                row.minutes = formattedMinutes
                table.addRow(row)
            }
        } catch {
            print("Failed to recognize minutes")
        }
    }
    
    return table
}

extension UIImage {
    func cropped(to rect: CGRect) -> UIImage? {
        guard let cgImage = self.cgImage?.cropping(to: rect) else {
            return nil
        }
        return UIImage(cgImage: cgImage)
    }
}
